components/collections/src/codepointtrie/cptrie.rs

```diff
@@ -861,6 +861,38 @@ impl<'trie, T: TrieValue> CodePointTrie<'trie, T> {
             .map(|cpm_range| cpm_range.range)
     }
 
+    /// Yields an [`Iterator`] returning the ranges of the code points whose values
+    /// match `predicate`.
+    ///
+    /// This is preferable to calling `.get_ranges().filter()` since it will coalesce
+    /// adjacent ranges into one.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_collections::codepointtrie::planes;
+    ///
+    /// let trie = planes::get_planes_trie();
+    ///
+    /// let plane_val = 2;
+    /// let mut sip_range_iter = trie.get_ranges_for_predicate(|r| r.value != plane_val as u8);
+    ///
+    /// let end = plane_val * 0x1_0000 - 1;
+    ///
+    /// let sip_range = sip_range_iter.next()
+    ///     .expect("Complemented planes data should have at least one entry");
+    /// assert_eq!(0..=end, sip_range);
+    pub fn get_ranges_for_predicate<'a>(
+        &'a self,
+        predicate: impl FnMut(&CodePointMapRange<T>) -> bool + Copy + 'a,
```

Comment:
question: should this predicate take an `&T` or an `&CodePointRange<T>`? I don't think the latter is too _useful_, it's just maximally permissive

Reply:
Strongly prefer the predicate input be `&T` unless a compelling reason exists otherwise. The `.get_range(u32)` method that powers `.iter_ranges()` iterates over code points.

That allows you to get performance improvements on `.get_range()` if you create a version that additionally takes a predicate fn, when you use that predicate fn to determine which code points to include in the current range that will be returned.